{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Ensamble de Genomas con ONT: Paso a paso**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simón Villanueva Corrales\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Los pasos generales para un ensamble con tecnología ONT son:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"https://raw.githubusercontent.com/sivico26/notebooks/master/Pipeline/Images/steps.png\" width=600>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Esta guía contiene una breve explicación de cada uno de los pasos y un caso sencillo que se utilizará para ilustrar el accionar de cada uno. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘Archivos_sim_ont’: File exists\n"
     ]
    }
   ],
   "source": [
    "mkdir Archivos_sim_ont"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Los programas que se utilizarán en cada paso destacan por su velocidad de procesamiento más que por su precisión u otra característica."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Simulación de datos**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Utilizaremos [Nanosim-h](https://github.com/karel-brinda/NanoSim-H), un simulador de datos de nanopore que ya está entrenado para generar reads de ciertos organismos. Generaremos datos, ¿cómo no?, de un genoma de *E. coli* de referencia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: nanosim-h [-h] [-p str] [-o str] [-n int] [-u float] [-m float]\n",
      "                 [-i float] [-d float] [-s int] [--circular] [--perfect]\n",
      "                 [--merge-contigs] [--rnf] [--rnf-add-cigar] [--max-len int]\n",
      "                 [--min-len int] [--kmer-bias int]\n",
      "                 <reference.fa>\n",
      "\n",
      "Program:  NanoSim-H - a simulator of Oxford Nanopore reads.\n",
      "Version:  1.1.0.3\n",
      "Authors:  Chen Yang <cheny@bcgsc.ca> - author of the original software package (NanoSim)\n",
      "          Karel Brinda <kbrinda@hsph.harvard.edu> - author of the NanoSim-H fork\n",
      "\n",
      "positional arguments:\n",
      "  <reference.fa>        reference genome (- for standard input)\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -p str, --profile str\n",
      "                        error profile - one of precomputed profiles\n",
      "                        ('ecoli_R7', 'ecoli_R9_2D', 'ecoli_R9_1D',\n",
      "                        'ecoli_UCSC1b', 'ecoli_R7.3', 'yeast') or own\n",
      "                        directory with an error profile [ecoli_R9_2D]\n",
      "  -o str, --out-pref str\n",
      "                        prefix of output file [simulated]\n",
      "  -n int, --number int  number of generated reads [10000]\n",
      "  -u float, --unalign-rate float\n",
      "                        rate of unaligned reads [detect from the error\n",
      "                        profile]\n",
      "  -m float, --mis-rate float\n",
      "                        mismatch rate (weight tuning) [1.0]\n",
      "  -i float, --ins-rate float\n",
      "                        insertion rate (weight tuning) [1.0]\n",
      "  -d float, --del-rate float\n",
      "                        deletion rate (weight tuning) [1.0]\n",
      "  -s int, --seed int    initial seed for the pseudorandom number generator (0\n",
      "                        for random) [42]\n",
      "  --circular            circular simulation (linear otherwise)\n",
      "  --perfect             output perfect reads, no mutations\n",
      "  --merge-contigs       merge contigs from the reference\n",
      "  --rnf                 use RNF format for read names\n",
      "  --rnf-add-cigar       add cigar to RNF names (not fully debugged, yet)\n",
      "  --max-len int         maximum read length [inf]\n",
      "  --min-len int         minimum read length [50]\n",
      "  --kmer-bias int       prohibits homopolymers with length >= n bases in\n",
      "                        output reads [6]\n",
      "\n",
      "Examples: nanosim-h --circular ecoli_ref.fasta\n",
      "          nanosim-h --circular --perfect ecoli_ref.fasta\n",
      "          nanosim-h -p yeast --kmer-bias 0 yeast_ref.fasta\n",
      "\n",
      "Notice: the use of `max-len` and `min-len` will affect the read length distributions. If\n",
      "the range between `max-len` and `min-len` is too small, the program will run slowlier accordingly.\n"
     ]
    }
   ],
   "source": [
    "!nanosim-h -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Archivos_sim_ont/sim_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "albacore_output   Ensamble\t Polishing\t\t      sim_reads\n",
      "e_coli_ref.fasta  Muestra_fast5  reads_lambda_albacore.fastq\n"
     ]
    }
   ],
   "source": [
    "!ls Archivos_sim_ont/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "albacore_output   Ensamble\t Polishing\t\t      sim_reads\n",
      "e_coli_ref.fasta  Muestra_fast5  reads_lambda_albacore.fastq\n"
     ]
    }
   ],
   "source": [
    "!ls ~/Documents/Notebooks/Pipeline/Archivos_sim_ont/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using profile ecoli_R9_2D from default profiles\n",
      "Simulating unaligned reads\n",
      "100% (6 of 6) |##########################| Elapsed Time: 0:00:00 Time:  0:00:00\n",
      "Simulating aligned reads\n",
      "100% (94 of 94) |########################| Elapsed Time: 0:00:02 Time:  0:00:02\n"
     ]
    }
   ],
   "source": [
    "!cd Archivos_sim_ont/sim_reads/ && nanosim-h --circular -n 100 --rnf -o ec ~/Documents/Notebooks/Pipeline/Archivos_sim_ont/e_coli_ref.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: nanosim-h [-h] [-p str] [-o str] [-n int] [-u float] [-m float]\n",
      "                 [-i float] [-d float] [-s int] [--circular] [--perfect]\n",
      "                 [--merge-contigs] [--rnf] [--rnf-add-cigar] [--max-len int]\n",
      "                 [--min-len int] [--kmer-bias int]\n",
      "                 <reference.fa>\n",
      "nanosim-h: error: argument <reference.fa>: can't open '/home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/e_coli.ref.fasta': [Errno 2] No such file or directory: '/home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/e_coli.ref.fasta'\n"
     ]
    }
   ],
   "source": [
    "!nanosim-h --circular -n 100 --rnf -o ecoli /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/e_coli.ref.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Basecalling**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Por la naturaleza de la tecnología (una buena explicación puede encontrarse en este [video](https://nanoporetech.com/resource-centre/videos/nanopore-dna-sequencing)), los datos que genera Nanopore son series de tiempo de señal eléctrica. Es por esto que lo primero que debe hacerse es interpretar esta señal y \"asignar las bases\"(basecall) correspondientes a cada variación."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"https://konradpaszkiewicz.files.wordpress.com/2014/04/capture5.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "El basecaller más utilizado actualmente es Albacore, desarrollado por ONT, el cual ofrece la mejor combinación entre precisión y velocidad."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Este paso es, en nuestro caso, el cuello de botella del proceso, ya que no podemos darnos el lujo de cambiar velocidad por precisión, puesto que afectaría seriamente los pasos posteriores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| 100 of 100|##############################################|100% Time:  0:00:30\n"
     ]
    }
   ],
   "source": [
    "!read_fast5_basecaller.py -r -i Archivos_sim_ont/Muestra_fast5/ -f FLO-MIN106 -k SQK-RAD002 -s Archivos_sim_ont/albacore_output -o fast5,fastq -t 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Verifiquemos cuantos de los reads evaluados pasaron"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "63\n"
     ]
    }
   ],
   "source": [
    "!ls Archivos_sim_ont/albacore_output/workspace/pass/0 | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "De un total de 100 reads de prueba, 63 pasaron el basecalling con suficiente calidad (usualmente es mayor la proporción)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Reads QC**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Antes de empezar con los siguientes pasos veamos los reads que tenemos (del fago lambda)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "El siguiente script hará un data frame de pandas para sacar los datos más relevantes de los reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "#Función para hacer un dataframe de pandas desde un fasta o fastq\n",
    "def read_sequence_file(filename, format):\n",
    "    rows = []\n",
    "    columns = ['id', 'sequence', 'read_length', 'gc_content']\n",
    "    if(format == 'fastq'):\n",
    "        columns.append('avg_quality')\n",
    "    for seq_record in SeqIO.parse(filename, format):\n",
    "        if(format == 'fastq'):\n",
    "            quality_score = np.mean(seq_record.letter_annotations[\"phred_quality\"])\n",
    "        read_id = seq_record.description\n",
    "        if(format == 'fastq'):\n",
    "            rows.append([read_id, seq_record.seq, len(seq_record.seq), GC(seq_record.seq), quality_score])\n",
    "        else:\n",
    "            rows.append([read_id, seq_record.seq, len(seq_record.seq), GC(seq_record.seq)])  \n",
    "        \n",
    "    df = pd.DataFrame(rows, columns=columns)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "seq_df = read_sequence_file('/home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/reads_lambda_albacore.fastq', 'fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_length</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>avg_quality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>23083.000000</td>\n",
       "      <td>23083.000000</td>\n",
       "      <td>23083.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>4795.898627</td>\n",
       "      <td>49.344589</td>\n",
       "      <td>13.678540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>6057.034726</td>\n",
       "      <td>5.959169</td>\n",
       "      <td>2.307329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>5.000000</td>\n",
       "      <td>17.241379</td>\n",
       "      <td>4.100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1461.000000</td>\n",
       "      <td>45.213929</td>\n",
       "      <td>12.097495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>3008.000000</td>\n",
       "      <td>48.780488</td>\n",
       "      <td>13.734980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>6035.000000</td>\n",
       "      <td>54.925265</td>\n",
       "      <td>15.275225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>193287.000000</td>\n",
       "      <td>71.949829</td>\n",
       "      <td>21.077437</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         read_length    gc_content   avg_quality\n",
       "count   23083.000000  23083.000000  23083.000000\n",
       "mean     4795.898627     49.344589     13.678540\n",
       "std      6057.034726      5.959169      2.307329\n",
       "min         5.000000     17.241379      4.100000\n",
       "25%      1461.000000     45.213929     12.097495\n",
       "50%      3008.000000     48.780488     13.734980\n",
       "75%      6035.000000     54.925265     15.275225\n",
       "max    193287.000000     71.949829     21.077437"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "¿Y eso cómo se ve? veamos unas distribuciones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhEAAAFrCAYAAABvz9soAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvNQv5yAAAIABJREFUeJzt3XmUZWV97vHvIy0ICnQjBi3ANApq0OuACCRmMGKYrtrEIAujoTUkrHvFJCZGhHgjHZVcEBOUpWIUEDCGUY2tF8WO4zVXBhlkRkpE6W4GpRmMODX+7h/7LT00Vd3Fps+prq7vZ61atc9vT+97dlX10+/e++xUFZIkSQ/Xo2a6AZIkaXYyREiSpF4MEZIkqRdDhCRJ6sUQIUmSejFESJKkXgwR0kYmyZIk/zrFvBclWT7qNrV9T9muKZavJDtPMe/VST6//lonqQ9DhDQCSW5J8uMk/5Xk9iSnJ3ncTLdrWIYdVqrqY1W1zzTacXqSdw6rHdJcZ4iQRudlVfU44LnA84CjZ7g9egSSzJvpNkgzzRAhjVhV3Q5cSBcmAEiyWZJ3J/lekjuSfDDJ5m3egiSfSfL9JHe36R0G1t0pyVeS/DDJMmDb6bYlyViSj7dtfyfJXw7MW5Lk3CRntm1fm2T3gfm7JbmizTsvyTlJ3pnkscBngbE28vJfScbaaptOtb0pvCTJTUnuSfL+JGn7fm2Sr7XpJDkxyZ1J7ktydZJnJTkceDVwZGvDp9vyv5Hky22b1yZ5+UCfHp/k0207l7b+fG1gfiU5IslNwE2t9t4kt7Z1LkvyO2u8h+cl+dfW56uTPC3J0a29tyZZ54iKtKEyREgj1gLA/sD4QPk44Gl0wWJnYHvgbW3eo4CPAL8OPBn4MfC+gXX/DbiMLjy8A1g8zXY8Cvg08M22v72BNybZd2CxlwNnA/OBpRP7TbIp8EngdGAb4CzgDwGq6ketfyur6nHta+XatrcWLwVeADwbOBjYd5Jl9gF+l+7927otd1dVfQj4GPCu1oaXJXl06/PngV8D/gL4WJKnt229H/gR8ES693Gy9/JAYE9g1/b6Urrjtg3dsTgvyWMGln8Z8FFgAXAFXYB8FN17/nbgX9bxHkgbLEOENDr/nuSHwK3AncAx0P1PGjgc+OuqWlVVPwT+ETgEoKruqqqPV9X9bd6xwO+1dZ9M94/s31fVT6vqq3T/SE7HC4AnVNXbq+pnVXUz8OGJ/TZfq6oLquoBun8In9PqewHzgJOq6udV9Qngkmnsc6rtTeW4qrqnqr4HfImB0ZsBPwe2BJ4BpKqur6rbptjeXsDj2nZ/VlVfBD4DvCrJJsAfAce09/o64IxJtvG/23H6MUBV/Ws7Rqur6p+AzYCnDyz/f6vqwqpaDZwHPKHt/+d0gWphkvnreB+kDZIhQhqdA6tqS+BFdP/gTZx2eAKwBXBZG2K/B/hcq5NkiyT/kuS7Se4DvgrMb//ojQF3t//9T/juNNvz63SnHO4Z2O/fAdsNLHP7wPT9wGPatQBjwIp68BP8bp3GPqfa3nSXf8jFqC0IvI9uFOHOJB9KstUU2xsDbq2qXwzUvks3KvAEumA02I/J+vSgWpK/TXJ9knvbe7g1Dz6ldMfA9I+BH7QQNfGayfolzQaGCGnEquordKcB3t1KP6D7x+SZVTW/fW3dLsIEeBPd/2z3rKqt6IbuAQLcBixo1yFMePI0m3Ir8J2Bfc6vqi2r6oBprHsbsP3ENQrNjoPdnGYb1ouqOqmqnk93iuFpwJunaMdKYMd2KmfCk4EVwPeB1cAOA/MG+/TL3U1MtOsfjqQ7hbKgquYD99IdG2mjZ4iQZsZ7gD9I8pz2v+IPAycm+TWAJNsPXJuwJV3IuCfJNrTTIABV9V3gG8A/JNk0yW/TnYOfjkuAHyZ5S5LNk2zSLkh8wTTW/TrwAPCGJPOSLAL2GJh/B/D4JFtPsy29JXlBkj3b9Q4/An4CTIw03AE8ZWDxi+lGNI5M8ugkL6J7v85uowOfAJa00Z9nAIeuY/db0gWP7wPzkrwNmGoURNroGCKkGVBV3wfO5FcXT76F7kLLi9opi//gV+fV3wNsTjdicRHdqY5Bf0x3od8quoBx5jTb8ADdhYvPBb7Ttn8K3XD8utb9GfAK4DDgHuA1dNcW/LTNv4HuYsub26mSsam2tR5sRRfC7qY7NXEXcEKbdyqwa2vDv7d2v4zuws8fAB8ADm3tBXgDXf9vp7tm46yJPk3hQrrj8a22758wvdM60kYhDz6lKUn9JLkY+GBVfWSm27K+JDkeeGJVTeuOF2mucSRCUi9Jfi/JE9vpjMV0t2GuOUoyqyR5RpJnt8+e2INupOWTM90uaUPlJ65J6uvpwLnAY4GbgYPWcmvlbLEl3SmMMbrrKf4J+NSMtkjagHk6Q5Ik9eLpDEmS1MucPp2x7bbb1sKFC2e6GZIkjcxll132g6p6wvrY1pwOEQsXLuQb3/jGTDdDkqSRSTLdT7VdJ09nSJKkXgwRkiSpF0OEJEnqxRAhSZJ6GVqISHJakjuTXDPJvDclqSTbttdJclKS8SRXJdltYNnFSW5qX4sH6s9PcnVb56Q1niYoSZKGbJgjEacD+61ZTLIjsA/wvYHy/sAu7etw4OS27MQTC/eke0LgMUkWtHVOBv58YL2H7EuSJA3P0EJEVX2V7qmCazoROBIY/KjMRcCZ1bkImJ/kScC+wLKqWlVVdwPLgP3avK2q6qLqPnLzTODAYfVFkiQ91EiviUiyCFhRVd9cY9b2PPjxuctbbW315ZPUJUnSiIwsRCTZAvg74G2j2ucU7VjSrseolStXzmRTJEma1UY5EvFUYCfgm0luAXYALk/yRGAFsOPAsju02trqO0xSX6eqWlJVqaqMjY317IokSRpZiKiqq6vq16pqYVUtpDsFsVtV3Q4sBQ5td2nsBdzbHil8IbBPkgXtgsp9gAvbvPuS7NXuyjgUH9crSdJIDfMWz7OArwNPT7I8yWFrWfwC4GZgHPgw8HqAqloFvAO4tH29vdVoy5zS1vk28Nlh9EOSJE0u3c0Nc9Puu+9ePoBLkjSXJLmsqnZfH9ua00/xXN/uOWFtgy2d+W8+dQQtkSRp+PzYa0mS1IshQpIk9WKIkCRJvRgiJElSL4YISZLUiyFCkiT1YoiQJEm9GCIkSVIvhghJktSLIUKSJPViiJAkSb0YIiRJUi+GCEmS1IshQpIk9WKIkCRJvRgiJElSL4YISZLUiyFCkiT1YoiQJEm9GCIkSVIvhghJktSLIUKSJPViiJAkSb0YIiRJUi+GCEmS1IshQpIk9WKIkCRJvRgiJElSL4YISZLUiyFCkiT1YoiQJEm9GCIkSVIvQwsRSU5LcmeSawZqJyS5IclVST6ZZP7AvKOTjCe5Mcm+A/X9Wm08yVED9Z2SXNzq5yTZdFh9kSRJDzXMkYjTgf3WqC0DnlVVzwa+BRwNkGRX4BDgmW2dDyTZJMkmwPuB/YFdgVe1ZQGOB06sqp2Bu4HDhtgXSZK0hqGFiKr6KrBqjdrnq2p1e3kRsEObXgScXVU/rarvAOPAHu1rvKpurqqfAWcDi5IEeDFwflv/DODAYfVFkiQ91ExeE/GnwGfb9PbArQPzlrfaVPXHA/cMBJKJ+jolWZKkktTKlSsfQfMlSZrbZiREJHkrsBr42Kj3XVVLqipVlbGxsVHvXpKkjca8Ue8wyWuBlwJ7V1W18gpgx4HFdmg1pqjfBcxPMq+NRgwuL0mSRmCkIxFJ9gOOBF5eVfcPzFoKHJJksyQ7AbsAlwCXAru0OzE2pbv4cmkLH18CDmrrLwY+Nap+SJKk4d7ieRbwdeDpSZYnOQx4H7AlsCzJlUk+CFBV1wLnAtcBnwOOqKoH2ijDG4ALgeuBc9uyAG8B/ibJON01EqcOqy+SJOmhhnY6o6peNUl5yn/oq+pY4NhJ6hcAF0xSv5nu7g1JkjQD/MRKSZLUiyFCkiT1YoiQJEm9GCIkSVIvhghJktSLIUKSJPViiJAkSb0YIiRJUi+GCEmS1IshQpIk9WKIkCRJvRgiJElSL4YISZLUiyFCkiT1YoiQJEm9GCIkSVIvhghJktSLIUKSJPViiJAkSb0YIiRJUi+GCEmS1IshQpIk9WKIkCRJvRgiJElSL4YISZLUiyFCkiT1YoiQJEm9GCIkSVIvhghJktSLIUKSJPViiJAkSb0YIiRJUi9DCxFJTktyZ5JrBmrbJFmW5Kb2fUGrJ8lJScaTXJVkt4F1Frflb0qyeKD+/CRXt3VOSpJh9UWSJD3UMEciTgf2W6N2FPCFqtoF+EJ7DbA/sEv7Ohw4GbrQARwD7AnsARwzETzaMn8+sN6a+5IkSUM0tBBRVV8FVq1RXgSc0abPAA4cqJ9ZnYuA+UmeBOwLLKuqVVV1N7AM2K/N26qqLqqqAs4c2JYkSRqBUV8TsV1V3dambwe2a9PbA7cOLLe81dZWXz5JfZ2SLElSSWrlypUPvweSJAmYwQsr2whCzcB+l1RVqipjY2Oj3r0kSRuNUYeIO9qpCNr3O1t9BbDjwHI7tNra6jtMUpckSSMy6hCxFJi4w2Ix8KmB+qHtLo29gHvbaY8LgX2SLGgXVO4DXNjm3Zdkr3ZXxqED25IkSSMwb1gbTnIW8CJg2yTL6e6yOA44N8lhwHeBg9viFwAHAOPA/cDrAKpqVZJ3AJe25d5eVRMXa76e7g6QzYHPti9JkjQiQwsRVfWqKWbtPcmyBRwxxXZOA06bpP4N4FmPpI2SJKk/P7FSkiT1YoiQJEm9GCIkSVIvhghJktSLIUKSJPViiJAkSb0YIiRJUi+GCEmS1IshQpIk9WKIkCRJvRgiJElSL4YISZLUiyFCkiT1YoiQJEm9GCIkSVIvhghJktSLIUKSJPViiJAkSb0YIiRJUi+GCEmS1IshQpIk9WKIkCRJvRgiJElSL4YISZLUiyFCkiT1YoiQJEm9GCIkSVIvhghJktSLIUKSJPViiJAkSb0YIiRJUi+GCEmS1IshQpIk9TIjISLJXye5Nsk1Sc5K8pgkOyW5OMl4knOSbNqW3ay9Hm/zFw5s5+hWvzHJvjPRF0mS5qqRh4gk2wN/CexeVc8CNgEOAY4HTqyqnYG7gcPaKocBd7f6iW05kuza1nsmsB/wgSSbjLIvkiTNZTN1OmMesHmSecAWwG3Ai4Hz2/wzgAPb9KL2mjZ/7yRp9bOr6qdV9R1gHNhjRO2XJGnOG3mIqKoVwLuB79GFh3uBy4B7qmp1W2w5sH2b3h64ta27ui3/+MH6JOtMKcmSJJWkVq5c+cg7JEnSHDUTpzMW0I0i7ASMAY+lOx0xElW1pKpSVRkbGxvVbiVJ2uhMK0QkefF0atP0EuA7VfX9qvo58AnghcD8dnoDYAdgRZteAezY9jkP2Bq4a7A+yTqSJGnIpjsS8e5p1qbje8BeSbZo1zbsDVwHfAk4qC2zGPhUm17aXtPmf7GqqtUPaXdv7ATsAlzSs02SJOlhmre2mUl2Bp4GbJXkgIFZW9NdEPmwVdXFSc4HLgdWA1cAHwL+D3B2kne22qltlVOBjyYZB1bR3ZFBVV2b5Fy6ALIaOKKqHujTJkmS9PCtNUTQnWZ4LbAd8OaB+n3Am/rutKqOAY5Zo3wzk9xdUVU/AV45xXaOBY7t2w5JktTfWkNEVZ0BnJHktVV1+miaJEmSZoN1jUQAUFWnJ3kq8NTBdarqgmE1TJIkbdimFSKS/CPw58D1wMR1BwUYIiRJmqOmFSKAg4GnVtV9w2yMJEmaPaZ7i+dtBghJkjRouiMRX09yFnAe8JOJotdESJI0d003RLygff+LgZrXREiSNIdN9+6M3x92QyRJ0uwy3bszDpis7ukMSZLmrumezhj8tMrHAM+l+9hqQ4QkSXNUr9MZSXblwcFCkiTNMdO9xfNBquo6YLf13BZJkjSL9Lkm4lF0d2v8fCgtkiRJs0KfayJWA+NM8WRNSZI0N3iLpyRJ6mW6pzMCHA68pJU+D5xSVTWshkmSpA3bdE9nvAt4HvCR9noxsAtw5DAaJUmSNnzTDRH7ArtV1WqAJOcCl2GIkCRpzpruLZ6he1bGhGo1SZI0R013JOJC4LNJTm+vF7eaJEmao9YaIpJsAmxGd9ricOAVbdZS4EPDbZokSdqQret0xnHAH1fVL6rqg1V1UFUdBPwEOHb4zZMkSRuqdYWIFwOnTVL/CDDpkz0lSdLcsK4QsUlV/WLNYqs9pC5JkuaOdYWIzZNssWYxyePorpWQJElz1LpCxDnAGUm2migk2Ro4BThvmA2TJEkbtnWFiLcDPwVWJLk8yeXAcuABYMmQ2yZJkjZga73Fs31C5WuS7Ez3sdcAV1TV+NBbJkmSNmjTfYrnON3jvyVJkoDpf+y1JEnSgxgiJElSL4YISZLUy4yEiCTzk5yf5IYk1yf5zSTbJFmW5Kb2fUFbNklOSjKe5Kokuw1sZ3Fb/qYki2eiL5IkzVUzNRLxXuBzVfUM4DnA9cBRwBeqahfgC+01wP7ALu3rcOBkgCTbAMcAewJ7AMdMBA9JkjR8Iw8R7cOqfhc4FaCqflZV9wCLgDPaYmcAB7bpRcCZ1bkImJ/kScC+wLKqWlVVdwPLgP1G2BVJkua0mRiJ2An4PvCRJFckOSXJY4Htquq2tsztwHZtenvg1oH1l7faVPW1SrIkSSWplStXPsKuSJI0d81EiJgH7AacXFXPA37Er05dAFBVBdQwdl5VS6oqVZWxsbFh7EKSpDlhJkLEcmB5VV3cXp9PFyruaKcpaN/vbPNXADsOrL9Dq01VlyRJIzDyEFFVtwO3Jnl6K+0NXAcsBSbusFgMfKpNLwUObXdp7AXc2057XAjsk2RBu6Byn1aTJEkjMK2PvR6CvwA+lmRT4GbgdXSB5twkhwHfBQ5uy14AHED3sdv3t2WpqlVJ3gFc2pZ7e1WtGl0XJEma22YkRFTVlcDuk8zae5JlCzhiiu2cBpy2flsnSZKmw0+slCRJvRgiJElSL4YISZLUiyFCkiT1YoiQJEm9GCIkSVIvhghJktSLIUKSJPViiJAkSb3M1Mdez1n3nHDYWufPf/OpI2qJJEmPjCMRkiSpF0OEJEnqxRAhSZJ6MURIkqReDBGSJKkXQ4QkSerFECFJknoxREiSpF4MEZIkqRdDhCRJ6sUQIUmSejFESJKkXgwRkiSpF0OEJEnqxRAhSZJ6MURIkqReDBGSJKkXQ4QkSerFECFJknoxREiSpF4MEZIkqRdDhCRJ6mXGQkSSTZJckeQz7fVOSS5OMp7knCSbtvpm7fV4m79wYBtHt/qNSfadmZ5IkjQ3zeRIxF8B1w+8Ph44sap2Bu4GDmv1w4C7W/3EthxJdgUOAZ4J7Ad8IMkmI2q7JElz3oyEiCQ7AP8dOKW9DvBi4Py2yBnAgW16UXtNm793W34RcHZV/bSqvgOMA3uMpgeSJGmmRiLeAxwJ/KK9fjxwT1Wtbq+XA9u36e2BWwHa/Hvb8r+sT7KOJEkaspGHiCQvBe6sqstGve+2/yVJKkmtXLlyJpogSdJGYSZGIl4IvDzJLcDZdKcx3gvMTzKvLbMDsKJNrwB2BGjztwbuGqxPss6UqmpJVaWqMjY29sh7I0nSHDXyEFFVR1fVDlW1kO7CyC9W1auBLwEHtcUWA59q00vba9r8L1ZVtfoh7e6NnYBdgEtG1A1Jkua8eeteZGTeApyd5J3AFcCprX4q8NEk48AquuBBVV2b5FzgOmA1cERVPTD6ZkuSNDfNaIioqi8DX27TNzPJ3RVV9RPglVOsfyxw7PBaKEmSpuInVkqSpF4MEZIkqRdDhCRJ6sUQIUmSejFESJKkXgwRkiSpF0OEJEnqxRAhSZJ6MURIkqReDBGSJKkXQ4QkSerFECFJknoxREiSpF4MEZIkqRdDhCRJ6sUQIUmSejFESJKkXgwRkiSpF0OEJEnqxRAhSZJ6MURIkqReDBGSJKkXQ4QkSerFECFJknoxREiSpF4MEZIkqRdDhCRJ6sUQIUmSejFESJKkXgwRkiSpF0OEJEnqxRAhSZJ6MURIkqReRh4ikuyY5EtJrktybZK/avVtkixLclP7vqDVk+SkJONJrkqy28C2Frflb0qyeNR9kSRpLps3A/tcDbypqi5PsiVwWZJlwGuBL1TVcUmOAo4C3gLsD+zSvvYETgb2TLINcAywO1BtO0ur6u6R92g9uueEw9Y6f/6bTx1RSyRJWruRj0RU1W1VdXmb/iFwPbA9sAg4oy12BnBgm14EnFmdi4D5SZ4E7Assq6pVLTgsA/YbYVckSZrTZvSaiCQLgecBFwPbVdVtbdbtwHZtenvg1oHVlrfaVPV17XNJkkpSK1eufETtlyRpLpuxEJHkccDHgTdW1X2D86qq6E5RrHdVtaSqUlUZGxsbxi4kSZoTZiREJHk0XYD4WFV9opXvaKcpaN/vbPUVwI4Dq+/QalPVJUnSCMzE3RkBTgWur6p/Hpi1FJi4w2Ix8KmB+qHtLo29gHvbaY8LgX2SLGh3cuzTapIkaQRm4u6MFwJ/Alyd5MpW+zvgOODcJIcB3wUObvMuAA4AxoH7gdcBVNWqJO8ALm3Lvb2qVo2mC5IkaeQhoqq+BmSK2XtPsnwBR0yxrdOA09Zf6yRJ0nT5iZWSJKkXQ4QkSerFECFJknoxREiSpF4MEZIkqRdDhCRJ6sUQIUmSejFESJKkXgwRkiSpF0OEJEnqxRAhSZJ6MURIkqReDBGSJKkXQ4QkSerFECFJknqZN9MN0MNzzwmHrXX+/DefOqKWSJLmOkciJElSL4YISZLUiyFCkiT1YoiQJEm9GCIkSVIvhghJktSLIUKSJPViiJAkSb0YIiRJUi+GCEmS1Isfe72R8WOxJUmj4kiEJEnqxRAhSZJ6MURIkqRevCZijvGaCUnS+uJIhCRJ6mXWj0Qk2Q94L7AJcEpVHTfDTZrVHKmQJE3XrB6JSLIJ8H5gf2BX4FVJdp3ZVkmSNDfM9pGIPYDxqroZIMnZwCLguhlt1UZsXSMV0+FohiRtHGZ7iNgeuHXg9XJgz7WtkGQJcEx7eX+S69dje8aAletxexuC9d+nI09br5t7mDxGG76NrT+w8fVpY+sPbHx9Wlt/fn197WS2h4iHraqWAEuGse0kVVVjw9j2TNnY+rSx9Qc2vj5tbP2Bja9PG1t/YOPr06j6M6uviQBWADsOvN6h1SRJ0pDN9hBxKbBLkp2SbAocAiyd4TZJkjQnzOrTGVW1OskbgAvpbvE8raquncEm/cMM7ntYNrY+bWz9gY2vTxtbf2Dj69PG1h/Y+Po0kv6kqkaxH0mStJGZ7aczJEnSDDFESJKkXgwRkiSpF0OEJEnqxRAhSZJ6MURIkqReDBHrSZL9ktyYZDzJUTPdnkFJdkzypSTXJbk2yV+1+pIkK5Jc2b4OGFjn6NaXG5PsO1CftJ/tA78ubvVz2od/DbtftyS5urX9G622TZJlSW5q3xe0epKc1Np3VZLdBrazuC1/U5LFA/Xnt+2Pt3UzxL48feA4XJnkviRvnG3HKMlpSe5Mcs1AbejHZKp9DKk/JyS5obX5k0nmt/rCJD8eOFYf7Nvutb03Q+rT0H/OkmzWXo+3+QuH2J9zBvpyS5IrW32DP0aZ+u/1hvl7VFV+PcIvug+6+jbwFGBT4JvArjPdroH2PQnYrU1vCXyL7tHpS4C/nWT5XVsfNgN2an3bZG39BM4FDmnTHwT+5wj6dQuw7Rq1dwFHtemjgOPb9AHAZ4EAewEXt/o2wM3t+4I2vaDNu6Qtm7bu/iP8ebqd7iE5s+oYAb8L7AZcM8pjMtU+htSffYB5bfr4gf4sHFxuje08rHZP9d4MsU9D/zkDXg98sE0fApwzrP6sMf+fgLfNlmPE1H+vN8jfI0ci1o9fPpK8qn4GTDySfINQVbdV1eVt+ofA9XRPQJ3KIuDsqvppVX0HGKfr46T9bCn2xcD5bf0zgAOH05t1WtT2v2Y7FgFnVuciYH6SJwH7AsuqalVV3Q0sA/Zr87aqqouq+406k9H1aW/g21X13bUss0Eeo6r6KrBqkrYO+5hMtY/13p+q+nxVrW4vL6J7Zs+UerZ7qvfmEZviGE1lff6cDfb1fGDvif8BD6s/bfsHA2etbRsb0jFay9/rDfL3yBCxfkz2SPK1/SM9Y9oQ4vOAi1vpDW0I7LSBoaup+jNV/fHAPQN/WEfV/wI+n+SyJIe32nZVdVubvh3Yrk0/3D5t36bXrI/CITz4j95sPkYwmmMy1T6G7U/p/ic3YackVyT5SpLfabU+7Z6JvynD/jn75Tpt/r1t+WH6HeCOqrppoDZrjtEaf683yN8jQ8QckuRxwMeBN1bVfcDJwFOB5wK30Q37zSa/XVW7AfsDRyT53cGZLWXPqs91b+ePXw6c10qz/Rg9yCiOyaiOe5K3AquBj7XSbcCTq+p5wN8A/5Zkq+lub4Z/Xjeqn7MBr+LBgXzWHKNJ/l6PtB3T3YchYv3Y4B9JnuTRdD+QH6uqTwBU1R1V9UBV/QL4MN0QJUzdn6nqd9ENoc1boz5UVbWifb8T+CRd+++YGFJs3+9siz/cPq3gwcPUozqm+wOXV9UdMPuPUTOKYzLVPoYiyWuBlwKvbn9saUP+d7Xpy+iuGXhaz3aP9G/KiH7OfrlOm791W34o2j5eAZwzUZstx2iyv9c92jGS3yNDxPqxQT+SvJ0XPBW4vqr+eaA+eP7uD4GJq5uXAoeku5p6J2AXugtxJu1n+yP6JeCgtv5i4FND7tNjk2w5MU13sds1re0TVyEPtmMpcGi7knkv4N42bHchsE+SBW0Idx/gwjbvviR7tffv0GH3qXnQ/5xm8zEaMIpjMtU+1rsk+wFHAi+vqvsH6k9IskmbfgrdMbm5Z7unem+G1adR/JwN9vUg4IsTAWxIXgLcUFW/HLqfDcdoqr/XPdoxmt+jWk9X/M71L7orZL9Fl2zfOtPtWaNtv003LHUVcGX7OgD4KHB1qy8FnjSwzltbX25k4K6EqfpJd5X2JXQXXp0HbDbkPj2F7orwbwLXTrSF7hzrF4CbgP8Atmn1AO9v7b4a2H1gW3/a2j0OvG6gvju922wWAAAD/UlEQVTdH9NvA++jPfV2iH16LN3/zLYeqM2qY0QXgG4Dfk53rvWwURyTqfYxpP6M051rnvhdmrjj4I/az+KVwOXAy/q2e23vzZD6NPSfM+Ax7fV4m/+UYfWn1U8H/scay27wx4ip/15vkL9HPgpckiT14ukMSZLUiyFCkiT1YoiQJEm9GCIkSVIvhghJktSLIUKahdI9mfCGJN9Mcn2SPxvCPr6c5KWT1F+U9tTUYUj3pMXD16jdkuRZ01j35UlO6LHPSfsqae0MEdLsdVBVPQd4JfCBJGMz3aD1ZCFw+LoWmkxVLa2qN6/f5kiaiiFCmuWq6hrgbgYe/pPkLUkuSXJ5kk8neWKr753k6+keQHR1kkMG1tk1ycVJrk1yNt2HA61TkgOS/Ge6B6F9vX1q3sSIxZVJ/iXdg52+meQ3BtY7Nsl42+fxA6Mb7wd2beueP7Crg9v2b0nyhina8tqJdda1/0m8pL1n40n+cWCbX07ynsnmSXOdIUKa5ZK8EPgB3ad3kuQ1dA9T2qu6B5RdwK8eqHQ53YPLnkf3scDvzq+e2PhR4ANV9UzgPcALprHvpwJ/T/dJhs8H/gw4d2CRZ9J9ouOzW/1/tfVeRvfsiecAv0n38cMTjgCuq6rnVtVBA/Utquo3gRcBx6V7QNG6TLr/KewK/BbdQ6hetsbpjbXNk+aseeteRNIG6vz22fc7A6+sqp+1+svpPtb28m428+geuwzwBOC0JLvQPYFyG+DpSa4DnkUXJKiqi5JcPY027EsXWL7a9gUwL8nEI4RvrKor2vRFwMva9O8D51bVjwCSnEEXRtbm7Na2W5LcTffgoBvWsc5U+5/MGdU9ovq/2kjMi4HPTGOeNGcZIqTZ66CquibJK4GPJPnP6p7+GeCdVXXaJOucTPdshFdUVSX5FtM8bTGFAJ+rqkMfMqM7dfCTgdIDPLK/OX22tT73L2kNns6QZrmqOg/4PHB0Ky0FXj9xmqI9gfE5bd584JYWIP6AbhSDqrqP7uE9f9zW2QP4b9PY/eeB/ZI8c6KQZJ2nQYAvAwcl2SLJo4A/GZh3H91jokftNUnmpXsq7MHAF6c5T5qzTOXSxuFo4LIkx1fVR5NsC3ylnWJ4FPABumsmjqK7k+Mf6B7nfNXANg6lG9E4ii5QXLqunVbVTe0ajFOTbA5sCvznutatqqVJfqvtfxXdqYaJazOuAm5Mcg3do5wPmmIz69sNwP+jO8VzblV9ZprzpDnLp3hKmhFJtqyqH7aRiFOAlVW1tgsfZ0SSLwPvNjhID+VIhKSZcmaShcDmwGXAu2a0NZIeNkciJElSL15YKUmSejFESJKkXgwRkiSpF0OEJEnqxRAhSZJ6+f/JW0UznlhQPQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fb3fc2bb080>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.mlab as mlab\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "\n",
    "#tamaño de la figura\n",
    "matplotlib.rcParams['figure.figsize'] = (9, 8)\n",
    "\n",
    "# ggplot style \n",
    "plt.style.use('seaborn-notebook')\n",
    "\n",
    "#On the plt.hist parameters we can set the number of bins and the bar color using labels as 'blue', 'red', 'yellow'... or using a RGB code.\n",
    "plt.hist(seq_df['read_length'], bins=50, color='#ef8a62')\n",
    "plt.title('Read length histogram')\n",
    "plt.xlabel('Read length in bp')\n",
    "plt.ylabel('Count')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "¿Y la calidad?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAgIAAAFrCAYAAACjVuHDAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvNQv5yAAAIABJREFUeJzt3Xm4ZFV57/HvTxAVBxpBkW6ITRSNaGIkiCQmiqKIhNjGoOJEo9z4eB0iavBiNKHjkEej0auJ0RhAGgcUiQNJ8GrHIcZEiICKDCotgnQ3g8ooOLW894+9jimOZ6L7VFXX2d/P89Rzdq299t7vrjrn1Ftrrb12qgpJktRPdxh3AJIkaXxMBCRJ6jETAUmSesxEQJKkHjMRkCSpx0wEJEnqMRMBSYsuSSW5f1t+d5I/X6T9rkny/jnWX5jkwMU4ltQX2487AGlSJPk88FDgPlX1kzGHMzGq6gVTy+1D+v1VtceQjvXg+eokWQl8B7hjVW0eRhzSJLFFQFqA9uHxe0ABTxrSMUzMlwDfR00aEwFpYY4EzgJOBlZPFSZ5RJKrkmw3UPaHSc5vy3dIclySbyf5QZLTktyzrVvZmtCPTvJd4LOt/CNtnzck+UKSBw/se5ck/5zkxiRfTvL6JF8cWP9rSdYluTbJN5M8bbYTSrJXkn9PclPb5u+mmt2THJhkw7T6lyV5XFveP8mXklyf5Mq27Q6zHOfkFuddgU8Cy5P8sD2WJ7klyS4D9fdN8r0kd5wl9B2SnNLivjDJfnPEeE57ra5O8tZW7Qvt5/Utht9u79Nrklye5Jq2/50G9ntkW/eDJH8+7Thrkpye5P1JbgSOmu/1ae/7C5Nc0s7jdUnul+S/WrynzfZ6SovNREBamCOBD7THE5LsBlBVZwM3A48dqPtM4INt+SXAk4FHA8uB64B3Ttv3o4EHAU9ozz8J7A3cGzivHXPKO9vx7kOXkAwmJXcF1rVj3xs4Avj7JPvMck4fBM4FdgVeN7ivBfg58LK27W8DBwEvnGuDqroZeCKwqaru1h6bgM8DgwnLc4APVdXPZtnVk4APAcuAM4C/m6Xe24G3V9U9gPsBp7XyR7Wfy1oMXwKOao/HAL8K3G1qv+31+3vgWcDuwE7AimnHWgWc3mL6AAt7fZ4A/BZwAPBK4D3As4E9gYcAz5jlvKRFZSIgzSPJ7wL3BU6rqnOBb9N92E85lfZPO8ndgUNbGcALgFdX1YY2rmANcPi05uM1VXVzVf0IoKpOqqqbBuo/NMlOrdXhj4Djq+qWqroIWDuwn8OAy6rqvVW1uaq+AvwT8NQZzulXgIcDf15VP6mqLwD/vNDXpKrOraqz2nEuA/6BLqHZEmvpPgBp5/gM4H1z1P9iVZ1ZVT9v9R46S72fAfdPsmtV/bCqzppjn88C3lpVl1bVD4FXAUe09+lw4J+r6otV9VPgL+i6iAZ9qao+XlW3VtWPFvj6/HVV3VhVFwIXAJ9ux7+BLhl82BzxSovGRECa32q6f9Lfb88/yG2/PX8QeEqSOwFPAc6rqsvbuvsCH2tNxNcDF9N9W9xtYPsrphaSbJfkja0r4UbgsrZqV+BedAN8r5hp23asR0wdqx3vWXStB9MtB65r39KnXD5DvRkleUCSf2ldGDcCf9Vi3BKfAPZJshfweOCGqvrvOepfNbB8C3DnzNwvfzTwAOAbrRvlsDn2uZzbnv/ldK/1bm3dL17nqroF+MG07Qffh4W+PlcPLP9ohud3myNeadE4qEWaQ5K70DVbb5dk6gPoTsCyJA+tqq9V1UVJLqdr9h7sFoDuA+J5VfWfM+x7ZVsc/Hb5TLpm5sfRJQE70XUnBPgesBnYA/hWq7/ntGP9e1U9fgGndiWwc5K7DiQDvzIQy83AjgOxbkeXiEx5F/AV4BlVdVOSY+i+Oc/nl253WlU/TnIaXavArzF3a8CCVdUlwDOS3IEuQTu9jUWY6Zarm+gSqSm/QvdaX033Wj1wakX7ndiF25q+zy19faSRs0VAmtuT6b7B7wP8Zns8CPgPunEDUz4IvJSu//kjA+XvBt6Q5L4ASe6VZNUcx7s78BO6b5w70n2TBKA1hX8UWJNkxyS/Ni2GfwEekOQ5Se7YHg9P8qDpB2ktFucAf5lkh9b98QcDVb5F903799ugvdfQJUCDcd4I/LDF8b/nOKdBVwO7DA7Ea06h66N/EouUCCR5dpJ7VdWtwPWt+Fa6hOpWurEAU04FXpZuAOXd6F73D7fLC08H/iDJ77QBfGvoErO5bOnrI42ciYA0t9XAe6vqu1V11dSDbiDZswaapE+l6wP+7EAXAnQD1s4APp3kJrorDx4xx/FOoWuW3ghc1OoPejFdK8FVdB+Yp9IlDlTVTcDBdIMEN7U6b+K2H+CDntliuRY4vh2btq8b6Aa3ndBiuRkYvIrgT9v2NwH/CHx4jnP6har6Rov50tZ9sbyV/yfdh/Ngt8rWOgS4MMkP6d6HI1r//S3AG4D/bDEcAJxE93p+gW6OgR/TDfSk9eG/hG6A4pXAD4FraK/7LLbo9ZHGIVUztZJJmgRJ3kQ3wdHtGfE/277WAPevqmdvdWBbdvzPAh+sqhPGcfyFai0G1wN7V9V3xh2PtLVsEZAmSLp5An4jnf3pBsR9bNxxba0kDwf2ZRv95pzkD1p3zF2BtwBf538GckoTzURAmix3pxsncDPdh+bf0I26n1hJ1gL/BhzTuje2Ravouls20c3xcETZnKolYmhdA0lOoruu+Zqqesi0da+gy6rvVVXfTxK6PrxD6S4HOqqqzmt1V9MNVAJ4fVUNXjctSZK2wjBbBE6mG6xzG0n2pBvQ9N2B4ifSZdl7A8+nu/SGdFOxHk83oGl/4PgkOw8xZkmSemVo8whU1RcGrpMe9Da66TQHmzNXAae0prazkixLsjtwILCuqq4FSLKOLrk4ldth1113rZUrZwpFkqSl59xzz/1+Vd1r/pojnlCoXT+9saq+1vUG/MIKbjsz14ZWNlv5Qo61hq41gd13351zzjlnywOXJGmCtEnOFmRkgwWT7Aj8Gd083UNXVWuqKlWV5cuXj+KQkiRNnFFeNXA/YC/ga0kuo5sm9bwk96GbsGRwqtQ9Wtls5ZIkaRGMLBGoqq9X1b2ramVVraRr5t+3zdJ2BnBkuzb6ALqbjlwJfAo4OMnObZDgwa1MkiQtgqElAklOBb4EPDDJhiRHz1H9TOBSYD3ddJwvBGiDBF8HfLk9Xjs1cFCSJG29XkwxvN9++5WDBSVJfZHk3KrabyF1nVlQkqQeMxGQJKnHTAQkSeoxEwFJknrMRECSpB4zEZAkqcdMBCRJ6jETAUmSemykdx+UpMV0/ZvnmrB0YZYde+IiRCJNLlsEJEnqMRMBSZJ6zERAkqQec4yApF6bb5yBYwi01NkiIElSj5kISJLUYyYCkiT1mImAJEk9ZiIgSVKPmQhIktRjXj4oaZu1GFMIS5qbLQKSJPWYiYAkST1mIiBJUo+ZCEiS1GMmApIk9ZiJgCRJPWYiIElSj5kISJLUYyYCkiT1mImAJEk95hTDksbGKYSl8bNFQJKkHjMRkCSpx4aWCCQ5Kck1SS4YKHtzkm8kOT/Jx5IsG1j3qiTrk3wzyRMGyg9pZeuTHDeseCVJ6qNhjhE4Gfg74JSBsnXAq6pqc5I3Aa8C/k+SfYAjgAcDy4F/S/KAts07gccDG4AvJzmjqi4aYtyS9AvzjWNYduyJI4pEGo6htQhU1ReAa6eVfbqqNrenZwF7tOVVwIeq6idV9R1gPbB/e6yvqkur6qfAh1pdSZK0CMY5RuB5wCfb8grgioF1G1rZbOXzSrImSSWpTZs2LUK4kiQtPWNJBJK8GtgMfGBYx6iqNVWVqsry5cuHdRhJkibayOcRSHIUcBhwUFVVK94I7DlQbY9WxhzlkiRpK400EUhyCPBK4NFVdcvAqjOADyZ5K91gwb2B/wYC7J1kL7oE4AjgmaOMWdKWc8Igads3tEQgyanAgcCuSTYAx9NdJXAnYF0SgLOq6gVVdWGS04CL6LoMXlRVP2/7eTHwKWA74KSqunBYMUvS7eVVBZp0Q0sEquoZMxTP+hdRVW8A3jBD+ZnAmYsYmiRJapxZUJKkHjMRkCSpx0wEJEnqMRMBSZJ6zERAkqQeMxGQJKnHTAQkSeoxEwFJknrMRECSpB4zEZAkqcdMBCRJ6jETAUmSemyktyGWtLR4m2Fp8tkiIElSj5kISJLUYyYCkiT1mImAJEk9ZiIgSVKPmQhIktRjJgKSJPWYiYAkST1mIiBJUo+ZCEiS1GMmApIk9ZiJgCRJPWYiIElSj5kISJLUYyYCkiT12PbjDkCSlrLr33z0nOuXHXviiCKRZmaLgCRJPWYiIElSj5kISJLUYyYCkiT1mImAJEk9NrREIMlJSa5JcsFA2T2TrEtySfu5cytPknckWZ/k/CT7DmyzutW/JMnqYcUrSVIfDbNF4GTgkGllxwGfqaq9gc+05wBPBPZuj+cD74IucQCOBx4B7A8cP5U8SJKkrTe0RKCqvgBcO614FbC2La8FnjxQfkp1zgKWJdkdeAKwrqqurarrgHX8cnIhSZK20KjHCOxWVVe25auA3dryCuCKgXobWtls5fNKsiZJJalNmzZtXdSSJC1RYxssWFUF1BD3v6aqUlVZvnz5sA4jSdJEG/UUw1cn2b2qrmxN/9e08o3AngP19mhlG4EDp5V/fgRxSmL+6XElTb5RtwicAUyN/F8NfGKg/Mh29cABwA2tC+FTwMFJdm6DBA9uZZIkaREMrUUgyal03+Z3TbKBbvT/G4HTkhwNXA48rVU/EzgUWA/cAjwXoKquTfI64Mut3muravoAREmStIWGlghU1TNmWXXQDHULeNEs+zkJOGkRQ5MkSY0zC0qS1GMmApIk9ZiJgCRJPWYiIElSj5kISJLUYyYCkiT1mImAJEk9ZiIgSVKPmQhIktRjJgKSJPWYiYAkST026tsQS5IGzHer52XHnjiiSNRXtghIktRjJgKSJPWYiYAkST1mIiBJUo+ZCEiS1GMmApIk9ZiJgCRJPWYiIElSj5kISJLUYyYCkiT1mImAJEk9ZiIgSVKPmQhIktRjJgKSJPWYiYAkST1mIiBJUo+ZCEiS1GMmApIk9dj24w5A0vhc/+ajxx2CpDGzRUCSpB4zEZAkqcdMBCRJ6rGxJAJJXpbkwiQXJDk1yZ2T7JXk7CTrk3w4yQ6t7p3a8/Vt/cpxxCxJ0lI08kQgyQrgT4D9quohwHbAEcCbgLdV1f2B64CpUUxHA9e18re1epIkaRGMq2tge+AuSbYHdgSuBB4LnN7WrwWe3JZXtee09QclyQhjlSRpyRp5IlBVG4G3AN+lSwBuAM4Frq+qza3aBmBFW14BXNG23dzq7zLfcZKsSVJJatOmTYt7EpIkLRHj6BrYme5b/l7AcuCuwCGLfZyqWlNVqaosX758sXcvSdKSMI6ugccB36mq71XVz4CPAo8ElrWuAoA9gI1teSOwJ0BbvxPwg9GGLEnS0jSOROC7wAFJdmx9/QcBFwGfAw5vdVYDn2jLZ7TntPWfraoaYbySJC1ZI59iuKrOTnI6cB6wGfgK8B7gX4EPJXl9KzuxbXIi8L4k64Fr6a4wkKRemG8a6GXHnjjnemk+Y7nXQFUdDxw/rfhSYP8Z6v4YeOoo4pIkqW+cWVCSpB5bUCKQ5LELKZMkSZNloS0Cb1lgmSRJmiBzjhFIcn/gAcA9khw6sGonuhkBJUnSBJtvsOAjgaOA3YBjB8pvBF4xpJgkSdKIzJkIVNVaYG2So6rq5NGEJGmxzHfpmSQt6PLBqjo5yf2A+w1uU1VnDiswSZI0fAtKBJL8FfDHwMXAz1txASYCkiRNsIVOKPQ04H5VdeMwg5EkSaO10MsHrzQJkCRp6Vloi8CXkpwKfAT48VShYwQkSZpsC00EHt5+vmSgzDECkiRNuIVeNfCYYQciSZJGb6FXDRw6U7ldA5IkTbaFdg0Mzip4Z+A3gfOwa0CSpIm2RV0DSfbhtsmBJEmaQAu9fPA2quoiYN9FjkWSJI3YlowRuAPdVQQ/G0pEkiRpZLZkjMBmYD3w1MUPR5IkjZKXD0qS1GML7RoI8Hzgca3o08AJVVXDCkySJA3fQrsG/hp4GPDe9nw1sDfwymEEJWlhrn/z0eMOQdKEW2gi8ARg36raDJDkNOBcTAQkSZpoC718MHT3FphSrUySJE2whbYIfAr4ZJKT2/PVrUySJE2wOROBJNsBd6LrAng+8JS26gzgPcMNTZIkDdt8XQNvBJ5ZVbdW1bur6vCqOhz4MfCG4YcnSZKGab5E4LHASTOUvxeY8Y6EkiRpcsw3RmC7qrp1emFV3Zrkl8olSaO1kEtIlx174ggi0aSar0XgLkl2nF6Y5G50YwckSdIEmy8R+DCwNsk9pgqS7AScAHxkmIFJkqThmy8ReC3wE2BjkvOSnAdsAH4OrBlybJIkacjmHCPQZhJ8dpL7000xDPCVqlo/9MgkSdLQLfTug+vpbj28KJIso+teeAjdLIXPA75J1xWxErgMeFpVXdduePR2uqsUbgGOqqrzFisWSZL6bKFTDC+2twP/r6p+DXgocDFwHPCZqtob+Ex7DvBEuhsc7U03qdG7Rh+uJElL08gTgTbY8FHAiQBV9dOquh5YBaxt1dYCT27Lq4BTqnMWsCzJ7iMOW5KkJWkcLQJ7Ad8D3pvkK0lOSHJXYLequrLVuQrYrS2vAK4Y2H5DK5tTkjVJKklt2rRpEcOXJGnpGEcisD2wL/CuqnoYcDP/0w0AQFUVt73b4e1WVWuqKlWV5cuXb82uJElassaRCGwANlTV2e356XSJwdVTTf7t5zVt/UZgz4Ht92hlkiRpK408Eaiqq4ArkjywFR0EXER3R8PVrWw18Im2fAZwZDoHADcMdCFIkqStsKDLB4fgJcAHkuwAXAo8ly4pOS3J0cDlwNNa3TPpLh1cT3f54HNHH64kSUvTWBKBqvoqsN8Mqw6aoW4BLxp6UJIk9dC4WgQkzWMhd5WTpK01rgmFJEnSNsAWAWlM/MYvaVtgi4AkST1mIiBJUo+ZCEiS1GMmApIk9ZiDBaUhcTCgpElgi4AkST1mIiBJUo+ZCEiS1GMmApIk9ZiJgCRJPWYiIElSj5kISJLUYyYCkiT1mBMKSdISN9/kVsuOPXFEkWhbZIuAJEk9ZiIgSVKPmQhIktRjJgKSJPWYiYAkST1mIiBJUo+ZCEiS1GMmApIk9ZgTCklbaL5JWiRpEtgiIElSj5kISJLUYyYCkiT1mImAJEk9ZiIgSVKPmQhIktRjJgKSJPWYiYAkST02tgmFkmwHnANsrKrDkuwFfAjYBTgXeE5V/TTJnYBTgN8CfgA8vaouG1PYkrTkzDc51rJjTxxRJBqHcbYIvBS4eOD5m4C3VdX9geuAqd/Mo4HrWvnbWj1JkrQIxpIIJNkD+H3ghPY8wGOB01uVtcCT2/Kq9py2/qBWX5IkbaVxtQj8X+CVwK3t+S7A9VW1uT3fAKxoyyuAKwDa+hta/TklWZOkktSmTZsWM3ZJkpaMkScCSQ4Drqmqc4d5nKpaU1WpqixfvnyYh5IkaWKNY7DgI4EnJTkUuDNwD+DtwLIk27dv/XsAG1v9jcCewIYk2wM70Q0alCRJW2nkLQJV9aqq2qOqVgJHAJ+tqmcBnwMOb9VWA59oy2e057T1n62qGmHIkiQtWdvSPAL/B3h5kvV0YwCmrlc5Edillb8cOG5M8UmStOSMbR4BgKr6PPD5tnwpsP8MdX4MPHWkgUmS1BPbUouAJEkaMRMBSZJ6zERAkqQeG+sYAUnSts97ESxttghIktRjtghIs5jvW5AkLQW2CEiS1GMmApIk9ZiJgCRJPWYiIElSj5kISJLUYyYCkiT1mImAJEk95jwCkqSt4syDk80WAUmSesxEQJKkHjMRkCSpx0wEJEnqMRMBSZJ6zERAkqQeMxGQJKnHTAQkSeoxEwFJknrMRECSpB4zEZAkqcdMBCRJ6jETAUmSesxEQJKkHvM2xJKkofI2xds2WwQkSeoxWwTUW/N9S5GkPrBFQJKkHjMRkCSpx0wEJEnqsZEnAkn2TPK5JBcluTDJS1v5PZOsS3JJ+7lzK0+SdyRZn+T8JPuOOmZJkpaqcbQIbAZeUVX7AAcAL0qyD3Ac8Jmq2hv4THsO8ERg7/Z4PvCu0YcsSdLSNPJEoKqurKrz2vJNwMXACmAVsLZVWws8uS2vAk6pzlnAsiS7jzhsSZKWpLGOEUiyEngYcDawW1Vd2VZdBezWllcAVwxstqGVzbfvNUkqSW3atGnRYpYkaSkZ2zwCSe4G/BNwTFXdmOQX66qqktTW7L+q1gBrAPbbb7+t2pckaXiceXC8xtIikOSOdEnAB6rqo6346qkm//bzmla+EdhzYPM9WpkkSdpK47hqIMCJwMVV9daBVWcAq9vyauATA+VHtqsHDgBuGOhCkCRJW2EcXQOPBJ4DfD3JV1vZnwFvBE5LcjRwOfC0tu5M4FBgPXAL8NzRhitJ0tI18kSgqr4IZJbVB81Qv4AXDTUoSZJ6ypkFJUnqMRMBSZJ6zERAkqQeMxGQJKnHTAQkSeoxEwFJknrMRECSpB4b270GpGGbb/5ySZItApIk9ZqJgCRJPWYiIElSj5kISJLUYyYCkiT1mFcNSJK2afNdAbTs2BNHFMnSZIuAJEk9ZiIgSVKPmQhIktRjJgKSJPWYiYAkST3mVQOSpInmVQVbx0RAE8ubCknS1rNrQJKkHjMRkCSpx0wEJEnqMRMBSZJ6zERAkqQeMxGQJKnHTAQkSeox5xGQJC1pC5lzpM+TDtkiIElSj5kISJLUYyYCkiT1mGMEtM3yXgKSRqXPNy6amEQgySHA24HtgBOq6o1jDklbwQ95Sdo2TEQikGQ74J3A44ENwJeTnFFVF403MklSHyzlFoOJSASA/YH1VXUpQJIPAasAE4FtlN/4JWkyTEoisAK4YuD5BuARc22QZA1wfHt6S5KLhxPaoloObBp3ECPUp/Pt07mC57vU9el8F3aurzxp+JHcPvddaMVJSQRut6paA6wZcxi3S5KqquXjjmNU+nS+fTpX8HyXuj6dbx/OdVIuH9wI7DnwfI9WJkmStsKkJAJfBvZOsleSHYAjgDPGHJMkSRNvIroGqmpzkhcDn6K7fPCkqrpwzGENw1+OO4AR69P59ulcwfNd6vp0vkv+XFNV445BkiSNyaR0DUiSpCEwEZAkqcdMBCRJ6jETAUmSesxEQJKkHjMRkCSpx0wExiDJZUm+nuSrSc6ZYX2SvCPJ+iTnJ9l3HHFurSQPbOc49bgxyTHT6hyY5IaBOn8xrni3RJKTklyT5IKBsnsmWZfkkvZz51m2Xd3qXJJk9eii3nKznO+bk3yj/a5+LMmyWbad8/d+WzTL+a5JsnHgd/bQWbY9JMk329/xcaOLesvMcq4fHjjPy5J8dZZtJ/G93TPJ55JclOTCJC9t5Uv273dWVeVjxA/gMmDXOdYfCnwSCHAAcPa4Y16Ec94OuAq477TyA4F/GXd8W3FejwL2BS4YKPtr4Li2fBzwphm2uydwafu5c1veedzns4XnezCwfVt+00zn29bN+Xu/LT5mOd81wJ/Os912wLeBXwV2AL4G7DPu87m95zpt/d8Af7GE3tvdgX3b8t2BbwH7LOW/39ketghsm1YBp1TnLGBZkt3HHdRWOgj4dlVdPu5AFlNVfQG4dlrxKmBtW14LPHmGTZ8ArKuqa6vqOmAdcMjQAl0kM51vVX26qja3p2fR3QtkSZjl/V2IX9w6vap+CkzdOn2bNde5JgnwNODUkQY1RFV1ZVWd15ZvAi6mu9Ptkv37nY2JwHgU8Okk5yZ5/gzrZ7rt8oqRRDY8RzD7P5HfTvK1JJ9M8uBRBjUku1XVlW35KmC3GeosxfcY4Hl0rVkzme/3fpK8uHWFnDRL0/FSe39/D7i6qi6ZZf1Ev7dJVgIPA86mh3+/JgLj8btVtS/wROBFSR417oCGqd0o6knAR2ZYfR5dd8FDgb8FPj7K2IatunbEXszjneTVwGbgA7NUWSq/9+8C7gf8JnAlXZP5UvcM5m4NmNj3NsndgH8CjqmqGwfX9eXv10RgDKpqY/t5DfAxumbEQUvttstPBM6rqqunr6iqG6vqh235TOCOSXYddYCL7Oqprpz285oZ6iyp9zjJUcBhwLPaP89fsoDf+4lQVVdX1c+r6lbgH5n5PJbM+5tke+ApwIdnqzOp722SO9IlAR+oqo+24t79/ZoIjFiSuya5+9Qy3UCrC6ZVOwM4sl09cABww0BT1SSa9dtEkvu0/keS7E/3O/mDEcY2DGcAU6OIVwOfmKHOp4CDk+zcmpYPbmUTJ8khwCuBJ1XVLbPUWcjv/USYNl7nD5n5PJbSrdMfB3yjqjbMtHJS39v2f+dE4OKqeuvAql79/QJeNTDqB90o4q+1x4XAq1v5C4AXtOUA76Qbdfx1YL9xx70V53tXug/2nQbKBs/1xe11+BrdQLPfGXfMt/P8TqVrHv4ZXT/h0cAuwGeAS4B/A+7Z6u4HnDCw7fOA9e3x3HGfy1ac73q6/tKvtse7W93lwJltecbf+239Mcv5vq/9XZ5P96Gx+/Tzbc8PpRuJ/u1JON+ZzrWVnzz19zpQdym8t79L1+x//sDv7qFL+e93toe3IZYkqcfsGpAkqcdMBCRJ6jETAUmSesxEQJKkHjMRkCSpx0wEpC3Qrh/+UZK3jzuWbUWSzyc5rC2/NsnT2/KBSQ4ewvHe2+4a90sT3SQ5Jsm9B56vSfKWxY5hsbS79z1k3HGon0wEpC3zTLp5D57RJoxZNG0mt4lWVX9RVVMf0AfSTbiyaJLsBvwR8OtV9fQZqhwD3HuG8sWMYeLfJwlMBKQt9Tzg9XSTkawCSLJjku8PTpGc5C1Jjm/Lj2j3Pz+3PX6/la9s270lyXnA/0pyUJIvJflKu8/7EQP73CfJ2UkuSPL+JGcNfBPfPcnpSf67bfdnMwWfZIck/9Dupf6lJH+f5PS27jbfngefzxXXtP2fnOTFSX6dbgKpI9Pdq/64JO9McuwUoapLAAAEjElEQVRA3Ycl+ebUDJPT9nNkO875ST6W5N5tFrvPATsC5yV52bRtXk034c3p7Zj7tFUrkpyZ5BtJ/jXJjgOvxZvba/a1JO9LN//89Fhmep9m3TbJM9v79JX2OGhgX7/XzuvrSf6ObhIxaTzGPaORDx+T9gB+g+7+6wGeBXxyYN0JwJ+05e2BTcBKYBnwFf5nFrrd6WZvW9bWF/D0gf3sDGzXlndrdXduz88Fnt2W9wN+DhzWnq8DHtWWdwD+A3j8DOfwEuDTwB3pPlDPAU5v69YAbxmo+4vn88T1+YE4TgZePMv+HkQ3G9vUhGYnAi+dIcaHtNdv6jV7HfDhtrwS+P4c79FlwEOmncMl7fVOO/c/buteA7xmoO6bgDfMsM+Z3qdZt6WboW7qHB8IbGjLd6Kbl/7A9vxpbb8Pme18fPgY5sOmLen2Oxo4paoqyUeBv02yorobr5wMvKM9nkg3R/tlSQ4F9gI+OfDFt4D7A98HfgycNnCMewEnJdmb7o5+9wQemOQiug/IDwJU1TlJzodfzPN+IHCvgWPcne6Dd920c3gMsLaqfgb8LMn76aZcnc+McdF1kyxIVV2c5FLgkCRn0d2Z8uUzVH0M3TS2U/fZ+Ae6aWy31Keq6nqAJGfT3UGQdvx7JDm8Pb/THMeZ/j7Nte39gFOTrKCbtvc+Se5D12VxS1V9HqCqTkvynq04L2mrmAhIt0MbD/BM4CdJjmzFdwSOovsm+MUkd29N4kcB753aFDi/qn7p9qzp7oV+c1UNzvf9Lrp57J/SEo5vAXceWD/T3OB3aOUPbx/wW2ozt+02HDzufHEt1DuAFwL7AB+tqhu2NNjb4ccDyz8H7tKWA7ywqj67gH1Mf5/m2vZU4BVV9fEkdwBuYfbXyrneNTaOEZBun1XAN6tqj6paWVUr6QbCHTVQZy3wCuBRdLc4BfgvurvRPWaqUpKHz9Qv3iwDLmsfto+nazmguvulX0h3R0eS7Av8elt3E11XwHEDx9izfQud7rPAc5Jsn+QudMnNlPXAbyW5Q+uPP2y+uOZxI7DTtLIz6VoSXk53g62ZfA44dCD+P+aXWzZuzzFncwbw8vY60BK5By3CtsuA77Tl59G1FgB8E7hLkt9r2xze6kpjYSIg3T7PAz4wWFBVXwLukOTRregU4DnAJ6rdlreqrqNrRj6+DSq7mK7ferZE4DjgLUm+SteHfP7AuiOBY5J8HfhTujvhTX2jfhawz9RANLp7yM/0IfMe4LvAxXRJwZcH1n0UuLat+yjdmISFxDWbjwEPnxosCFBVt9IlTN+pqhn3UVUXtOOta90fDwVeuoDjQdfi8N5pgwVn80a65vwvt+N8ka47ZSHm2vYY4ONtYOGv0m6vXVU/oUvk/r5tcyDdeyGNhXcflCZMG5V+c/tWvg/dIL0HtmRjS/d5FN1Av8Pnq7tYkqwD3lNVHxnVMSX9MlsEpMnzO8BX27fJD9GNft/iJGDUkuyX5Nt0rRj/NF99ScNli4AkST1mi4AkST1mIiBJUo+ZCEiS1GMmApIk9ZiJgCRJPfb/Af2ceKE65tF7AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fb3fa0c5438>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Only works with FASTQ files\n",
    "plt.hist(seq_df['avg_quality'], bins=50, color='#ef8a62')\n",
    "plt.title('Average quality histogram')\n",
    "plt.xlabel('Average quality of the read')\n",
    "plt.ylabel('Count')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Read to Read Overlap**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ya que tenemos los reads con sus bases asignadas (formato fastq o fasta), seguimos con el overlap, que es un mapeo buscando coincidencias entre los reads para posteriormente pasar al ensamble. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "De aquí en más usaremos datos del fago lambda"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "El programa que usaremos será [Minimap2](https://github.com/lh3/minimap2), la versión mejorada de Minimap. Minimap2 cuenta con una configuración precargada especificamente para este paso: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: minimap2 [options] <target.fa>|<target.idx> [query.fa] [...]\n",
      "Options:\n",
      "  Indexing:\n",
      "    -H           use homopolymer-compressed k-mer\n",
      "    -k INT       k-mer size (no larger than 28) [15]\n",
      "    -w INT       minizer window size [10]\n",
      "    -I NUM       split index for every ~NUM input bases [4G]\n",
      "    -d FILE      dump index to FILE []\n",
      "  Mapping:\n",
      "    -f FLOAT     filter out top FLOAT fraction of repetitive minimizers [0.0002]\n",
      "    -g NUM       stop chain enlongation if there are no minimizers in INT-bp [5000]\n",
      "    -G NUM       max intron length (effective with -xsplice; changing -r) [200k]\n",
      "    -F NUM       max fragment length (effective with -xsr or in the fragment mode) [800]\n",
      "    -r NUM       bandwidth used in chaining and DP-based alignment [500]\n",
      "    -n INT       minimal number of minimizers on a chain [3]\n",
      "    -m INT       minimal chaining score (matching bases minus log gap penalty) [40]\n",
      "    -X           skip self and dual mappings (for the all-vs-all mode)\n",
      "    -p FLOAT     min secondary-to-primary score ratio [0.8]\n",
      "    -N INT       retain at most INT secondary alignments [5]\n",
      "  Alignment:\n",
      "    -A INT       matching score [2]\n",
      "    -B INT       mismatch penalty [4]\n",
      "    -O INT[,INT] gap open penalty [4,24]\n",
      "    -E INT[,INT] gap extension penalty; a k-long gap costs min{O1+k*E1,O2+k*E2} [2,1]\n",
      "    -z INT[,INT] Z-drop score and inversion Z-drop score [400,200]\n",
      "    -s INT       minimal peak DP alignment score [80]\n",
      "    -u CHAR      how to find GT-AG. f:transcript strand, b:both strands, n:don't match GT-AG [n]\n",
      "  Input/Output:\n",
      "    -a           output in the SAM format (PAF by default)\n",
      "    -Q           don't output base quality in SAM\n",
      "    -L           write CIGAR with >65535 ops at the CG tag\n",
      "    -R STR       SAM read group line in a format like '@RG\\tID:foo\\tSM:bar' []\n",
      "    -c           output CIGAR in PAF\n",
      "    --cs[=STR]   output the cs tag; STR is 'short' (if absent) or 'long' [none]\n",
      "    --MD         output the MD tag\n",
      "    -Y           use soft clipping for supplementary alignments\n",
      "    -t INT       number of threads [3]\n",
      "    -K NUM       minibatch size for mapping [500M]\n",
      "    --version    show version number\n",
      "  Preset:\n",
      "    -x STR       preset (always applied before other options) []\n",
      "                 map-pb: -Hk19 (PacBio vs reference mapping)\n",
      "                 map-ont: -k15 (Oxford Nanopore vs reference mapping)\n",
      "                 asm5: -k19 -w19 -A1 -B19 -O39,81 -E3,1 -s200 -z200 (asm to ref mapping; break at 5% div.)\n",
      "                 asm10: -k19 -w19 -A1 -B9 -O16,41 -E2,1 -s200 -z200 (asm to ref mapping; break at 10% div.)\n",
      "                 ava-pb: -Hk19 -Xw5 -m100 -g10000 --max-chain-skip 25 (PacBio read overlap)\n",
      "                 ava-ont: -k15 -Xw5 -m100 -g10000 -r2000 --max-chain-skip 25 (ONT read overlap)\n",
      "                 splice: long-read spliced alignment (see minimap2.1 for details)\n",
      "                 sr: short single-end reads without splicing (see minimap2.1 for details)\n",
      "\n",
      "See `man ./minimap2.1' for detailed description of command-line options.\n"
     ]
    }
   ],
   "source": [
    "! minimap2 -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Archivos_sim_ont/Ensamble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::5.540*1.30] collected minimizers\n",
      "[M::mm_idx_gen::7.445*1.73] sorted minimizers\n",
      "[M::main::7.445*1.73] loaded/built the index for 23083 target sequence(s)\n",
      "[M::mm_mapopt_update::8.096*1.67] mid_occ = 401\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 5; is_hpc: 0; #seq: 23083\n",
      "[M::mm_idx_stat::8.503*1.64] distinct minimizers: 19092294 (80.61% are singletons); average occurrences: 1.968; average spacing: 2.947\n",
      "[M::worker_pipeline::168.868*2.64] mapped 23083 sequences\n",
      "[M::main] Version: 2.10-r761\n",
      "[M::main] CMD: minimap2 -x ava-ont Archivos_sim_ont/reads_lambda_albacore.fastq Archivos_sim_ont/reads_lambda_albacore.fastq\n",
      "[M::main] Real time: 168.905 sec; CPU: 445.987 sec\n"
     ]
    }
   ],
   "source": [
    "!minimap2 -x ava-ont Archivos_sim_ont/reads_lambda_albacore.fastq Archivos_sim_ont/reads_lambda_albacore.fastq > Archivos_sim_ont/Ensamble/ovlp_fl.paf "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Ensamble**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Una vez hallados los overlaps, se procedeal ensamble como tal, en el que se toman los averlaps y se trata de construir en el menor número de contigs posibles (idealmente el número de contigs es igual al número de cromosomas)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Usaremos [Miniasm](https://github.com/lh3/miniasm), un ensamblador OLC desarrollado para ir de la mano con minimap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "miniasm: option requires an argument -- 'h'\n",
      "Usage: miniasm [options] <in.paf>\n",
      "Options:\n",
      "  Pre-selection:\n",
      "    -m INT      min match length [100]\n",
      "    -i FLOAT    min identity [0.05]\n",
      "    -s INT      min span [2000]\n",
      "    -c INT      min coverage [3]\n",
      "  Overlap:\n",
      "    -o INT      min overlap [same as -s]\n",
      "    -h INT      max over hang length [1000]\n",
      "    -I FLOAT    min end-to-end match ratio [0.8]\n",
      "  Layout:\n",
      "    -g INT      max gap differences between reads for trans-reduction [1000]\n",
      "    -d INT      max distance for bubble popping [50000]\n",
      "    -e INT      small unitig threshold [4]\n",
      "    -f FILE     read sequences []\n",
      "    -n INT      rounds of short overlap removal [3]\n",
      "    -r FLOAT[,FLOAT]\n",
      "                max and min overlap drop ratio [0.7,0.5]\n",
      "    -F FLOAT    aggressive overlap drop ratio in the end [0.8]\n",
      "  Miscellaneous:\n",
      "    -p STR      output information: bed, paf, sg or ug [ug]\n",
      "    -b          both directions of an arc are present in input\n",
      "    -1          skip 1-pass read selection\n",
      "    -2          skip 2-pass read selection\n",
      "    -V          print version number\n",
      "\n",
      "See miniasm.1 for detailed description of the command-line options.\n"
     ]
    }
   ],
   "source": [
    "!miniasm -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::main] ===> Step 1: reading read mappings <===\n",
      "[M::ma_hit_read::25.667*1.00] read 13671141 hits; stored 17996729 hits and 13331 sequences (93372637 bp)\n",
      "[M::main] ===> Step 2: 1-pass (crude) read selection <===\n",
      "[M::ma_hit_sub::30.126*1.00] 12950 query sequences remain after sub\n",
      "[M::ma_hit_cut::30.540*1.00] 17912211 hits remain after cut\n",
      "[M::ma_hit_flt::31.055*1.00] 17110407 hits remain after filtering; crude coverage after filtering: 1015.33\n",
      "[M::main] ===> Step 3: 2-pass (fine) read selection <===\n",
      "[M::ma_hit_sub::33.229*1.00] 12906 query sequences remain after sub\n",
      "[M::ma_hit_cut::33.629*1.00] 17084609 hits remain after cut\n",
      "[M::ma_hit_contained::34.204*1.00] 16 sequences and 156 hits remain after containment removal\n",
      "[M::main] ===> Step 4: graph cleaning <===\n",
      "[M::ma_sg_gen] read 138 arcs\n",
      "[M::main] ===> Step 4.1: transitive reduction <===\n",
      "[M::asg_arc_del_trans] transitively reduced 97 arcs\n",
      "[M::asg_arc_del_multi] removed 0 multi-arcs\n",
      "[M::asg_arc_del_asymm] removed 3 asymmetric arcs\n",
      "[M::main] ===> Step 4.2: initial tip cutting and bubble popping <===\n",
      "[M::asg_cut_tip] cut 4 tips\n",
      "[M::asg_pop_bubble] popped 1 bubbles and trimmed 0 tips\n",
      "[M::main] ===> Step 4.3: cutting short overlaps (3 rounds in total) <===\n",
      "[M::asg_arc_del_short] removed 0 short overlaps\n",
      "[M::asg_arc_del_short] removed 0 short overlaps\n",
      "[M::asg_arc_del_short] removed 0 short overlaps\n",
      "[M::main] ===> Step 4.4: removing short internal sequences and bi-loops <===\n",
      "[M::asg_cut_internal] cut 0 internal sequences\n",
      "[M::asg_cut_biloop] cut 0 small bi-loops\n",
      "[M::asg_cut_tip] cut 0 tips\n",
      "[M::asg_pop_bubble] popped 0 bubbles and trimmed 0 tips\n",
      "[M::main] ===> Step 4.5: aggressively cutting short overlaps <===\n",
      "[M::asg_arc_del_short] removed 0 short overlaps\n",
      "[M::main] ===> Step 5: generating unitigs <===\n",
      "[M::main] Version: 0.2-r128\n",
      "[M::main] CMD: miniasm -f Archivos_sim_ont/reads_lambda_albacore.fastq Archivos_sim_ont/Ensamble/ovlp_fl.paf\n",
      "[M::main] Real time: 34.631 sec; CPU: 34.632 sec\n"
     ]
    }
   ],
   "source": [
    "!miniasm -f Archivos_sim_ont/reads_lambda_albacore.fastq Archivos_sim_ont/Ensamble/ovlp_fl.paf > Archivos_sim_ont/Ensamble/fago_lambda_raw.gfa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ahora cambiemos el formato a algo más conocido (gfa to fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "!awk '/^S/{print \">\"$2\"\\n\"$3}' Archivos_sim_ont/Ensamble/fago_lambda_raw.gfa | fold > Archivos_sim_ont/Ensamble/fago_lambda_raw.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ahora veamos el número de contigs del ensamble:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "!grep -c \">\" Archivos_sim_ont/Ensamble/fago_lambda_raw.fasta  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "¡Un exitoso ensamble de un solo Contig!\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Read Mapping**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aunque hayamos logrado hacer un ensamble de un solo contig, es bastante probable que esté lleno de errores, por lo que hay que corregirlos. El primer paso en para hacer esto es volver a mappear los reads en al ensamble."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Para esto, volveremos a utilizar Minimap2, pero con otra configuración"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Archivos_sim_ont/Polishing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::0.005*1.30] collected minimizers\n",
      "[M::mm_idx_gen::0.009*1.94] sorted minimizers\n",
      "[M::main::0.009*1.93] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.009*1.84] mid_occ = 2\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.010*1.76] distinct minimizers: 9304 (100.00% are singletons); average occurrences: 1.000; average spacing: 5.324\n",
      "[M::worker_pipeline::2.995*2.52] mapped 23083 sequences\n",
      "[M::main] Version: 2.10-r761\n",
      "[M::main] CMD: minimap2 -x map-ont Archivos_sim_ont/Ensamble/fago_lambda_raw.fasta Archivos_sim_ont/reads_lambda_albacore.fastq\n",
      "[M::main] Real time: 2.997 sec; CPU: 7.561 sec\n"
     ]
    }
   ],
   "source": [
    "!minimap2 -x map-ont Archivos_sim_ont/Ensamble/fago_lambda_raw.fasta Archivos_sim_ont/reads_lambda_albacore.fastq > Archivos_sim_ont/Polishing/aln_fl.paf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Polishing**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finalmente haremos polishing, que consiste en corregir el ensamble que tenemos al compararlo nuevamente con los reads que fueron mapeados en el paso anterior"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Para un polishing rápido usaremos [Racon](https://github.com/isovic/racon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!racon /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/reads_lambda_albacore.fastq /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/Polishing/aln_fl.paf /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/Ensamble/fago_lambda_raw.fasta > /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/Polishing/fago_lambda_pol.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Por alguna razón el resultado no se muestra, pero `Racon` efectivamente corrió"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Assembly QC**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "¿Cómo de determinar si el ensamble es realmente bueno? suelen usarse varias métricas para evaluar el ensamble, como el número de contigs, el n50, etc; pero también se debe evaluar a precisión y para eso se debe comparar con una referencia. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QUAST: Quality Assessment Tool for Genome Assemblies\n",
      "Version: 5.0.0.dev0, 1563de2\n",
      "\n",
      "Usage: python /home/sivico26/anaconda3/bin/quast.py [options] <files_with_contigs>\n",
      "\n",
      "Options:\n",
      "-o  --output-dir  <dirname>       Directory to store all result files [default: quast_results/results_<datetime>]\n",
      "-r                <filename>      Reference genome file\n",
      "-g  --features [type:]<filename>  File with genomic feature coordinates in the reference (GFF, BED, NCBI or TXT)\n",
      "                                  Optional 'type' can be specified for extracting only a specific feature type from GFF\n",
      "-m  --min-contig  <int>           Lower threshold for contig length [default: 500]\n",
      "-t  --threads     <int>           Maximum number of threads [default: 25% of CPUs]\n",
      "\n",
      "Advanced options:\n",
      "-s  --scaffolds                       Assemblies are scaffolds, split them and add contigs to the comparison\n",
      "-l  --labels \"label, label, ...\"      Names of assemblies to use in reports, comma-separated. If contain spaces, use quotes\n",
      "-L                                    Take assembly names from their parent directory names\n",
      "-e  --eukaryote                       Genome is eukaryotic (primarily affects gene prediction)\n",
      "    --fungus                          Genome is fungal (primarily affects gene prediction)\n",
      "    --large                           Use optimal parameters for evaluation of large genomes\n",
      "                                      In particular, imposes '-e -m 3000 -i 500 -x 7000' (can be overridden manually)\n",
      "-k  --k-mer-stats                     Compute k-mer-based quality metrics (recommended for large genomes)\n",
      "                                      This may significantly increase memory and time consumption on large genomes\n",
      "    --k-mer-size                      Size of k used in --k-mer-stats [default: 101]\n",
      "    --circos                          Draw Circos plot\n",
      "-f  --gene-finding                    Predict genes using GeneMarkS (prokaryotes, default) or GeneMark-ES (eukaryotes, use --eukaryote)\n",
      "    --mgm                             Use MetaGeneMark for gene prediction (instead of the default finder, see above)\n",
      "    --glimmer                         Use GlimmerHMM for gene prediction (instead of the default finder, see above)\n",
      "    --gene-thresholds <int,int,...>   Comma-separated list of threshold lengths of genes to search with Gene Finding module\n",
      "                                      [default: 0,300,1500,3000]\n",
      "    --rna-finding                     Predict ribosomal RNA genes using Barrnap\n",
      "-b  --conserved-genes-finding         Count conserved orthologs using BUSCO (only on Linux)\n",
      "-O  --operons  <filename>             File with operon coordinates in the reference (GFF, BED, NCBI or TXT)\n",
      "    --est-ref-size <int>              Estimated reference size (for computing NGx metrics without a reference)\n",
      "    --contig-thresholds <int,int,...> Comma-separated list of contig length thresholds [default: 0,1000,5000,10000,25000,50000]\n",
      "-u  --use-all-alignments              Compute genome fraction, # genes, # operons in QUAST v1.* style.\n",
      "                                      By default, QUAST filters Minimap's alignments to keep only best ones\n",
      "-i  --min-alignment <int>             The minimum alignment length [default: 65]\n",
      "    --min-identity <float>            The minimum alignment identity (80.0, 100.0) [default: 95.0]\n",
      "-a  --ambiguity-usage <none|one|all>  Use none, one, or all alignments of a contig when all of them\n",
      "                                      are almost equally good (see --ambiguity-score) [default: one]\n",
      "    --ambiguity-score <float>         Score S for defining equally good alignments of a single contig. All alignments are sorted \n",
      "                                      by decreasing LEN * IDY% value. All alignments with LEN * IDY% < S * best(LEN * IDY%) are \n",
      "                                      discarded. S should be between 0.8 and 1.0 [default: 0.99]\n",
      "    --strict-NA                       Break contigs in any misassembly event when compute NAx and NGAx.\n",
      "                                      By default, QUAST breaks contigs only by extensive misassemblies (not local ones)\n",
      "-x  --extensive-mis-size  <int>       Lower threshold for extensive misassembly size. All relocations with inconsistency\n",
      "                                      less than extensive-mis-size are counted as local misassemblies [default: 1000]\n",
      "    --scaffold-gap-max-size  <int>    Max allowed scaffold gap length difference. All relocations with inconsistency\n",
      "                                      less than scaffold-gap-size are counted as scaffold gap misassemblies [default: 10000]\n",
      "    --unaligned-part-size  <int>      Lower threshold for detecting partially unaligned contigs. Such contig should have\n",
      "                                      at least one unaligned fragment >= the threshold [default: 500]\n",
      "    --fragmented                      Reference genome may be fragmented into small pieces (e.g. scaffolded reference) \n",
      "    --fragmented-max-indent  <int>    Mark translocation as fake if both alignments are located no further than N bases \n",
      "                                      from the ends of the reference fragments [default: 50]\n",
      "                                      Requires --fragmented option\n",
      "    --upper-bound-assembly            Simulate upper bound assembly based on the reference genome and reads\n",
      "    --upper-bound-min-con  <int>      Minimal number of 'connecting reads' needed for joining upper bound contigs into a scaffold\n",
      "                                      [default: 2 for mate-pairs and 1 for long reads]\n",
      "    --est-insert-size  <int>          Use provided insert size in upper bound assembly simulation [default: auto detect from reads or 255]\n",
      "    --plots-format  <str>             Save plots in specified format [default: pdf].\n",
      "                                      Supported formats: emf, eps, pdf, png, ps, raw, rgba, svg, svgz\n",
      "    --memory-efficient                Run everything using one thread, separately per each assembly.\n",
      "                                      This may significantly reduce memory consumption on large genomes\n",
      "    --space-efficient                 Create only reports and plots files. Aux files including .stdout, .stderr, .coords will not be created.\n",
      "                                      This may significantly reduce space consumption on large genomes. Icarus viewers also will not be built\n",
      "-1  --reads1  <filename>              File with forward reads (in FASTQ format, may be gzipped)\n",
      "-2  --reads2  <filename>              File with reverse reads (in FASTQ format, may be gzipped)\n",
      "    --12      <filename>              File with interlaced forward and reverse paired-end reads. (in FASTQ format, may be gzipped)\n",
      "    --pe1     <filename>              File with forward paired-end reads (in FASTQ format, may be gzipped)\n",
      "    --pe2     <filename>              File with reverse paired-end reads (in FASTQ format, may be gzipped)\n",
      "    --pe12    <filename>              File with interlaced forward and reverse paired-end reads (in FASTQ format, may be gzipped)\n",
      "    --mp1     <filename>              File with forward mate-pair reads (in FASTQ format, may be gzipped)\n",
      "    --mp2     <filename>              File with reverse mate-pair reads (in FASTQ format, may be gzipped)\n",
      "    --mp12    <filename>              File with interlaced forward and reverse mate-pair reads (in FASTQ format, may be gzipped)\n",
      "    --single  <filename>              File with unpaired reads (in FASTQ format, may be gzipped)\n",
      "    --pacbio  <filename>              File with PacBio reads (in FASTQ format, may be gzipped)\n",
      "    --nanopore  <filename>            File with Oxford Nanopore reads (in FASTQ format, may be gzipped)\n",
      "    --ref-sam <filename>              SAM alignment file obtained by aligning reads to reference genome file\n",
      "    --ref-bam <filename>              BAM alignment file obtained by aligning reads to reference genome file\n",
      "    --sam     <filename,filename,...> Comma-separated list of SAM alignment files obtained by aligning reads to assemblies\n",
      "                                      (use the same order as for files with contigs)\n",
      "    --bam     <filename,filename,...> Comma-separated list of BAM alignment files obtained by aligning reads to assemblies\n",
      "                                      (use the same order as for files with contigs)\n",
      "                                      Reads (or SAM/BAM file) are used for structural variation detection and\n",
      "                                      coverage histogram building in Icarus\n",
      "    --sv-bedpe  <filename>            File with structural variations (in BEDPE format)\n",
      "\n",
      "Speedup options:\n",
      "    --no-check                        Do not check and correct input fasta files. Use at your own risk (see manual)\n",
      "    --no-plots                        Do not draw plots\n",
      "    --no-html                         Do not build html reports and Icarus viewers\n",
      "    --no-icarus                       Do not build Icarus viewers\n",
      "    --no-snps                         Do not report SNPs (may significantly reduce memory consumption on large genomes)\n",
      "    --no-gc                           Do not compute GC% and GC-distribution\n",
      "    --no-sv                           Do not run structural variation detection (make sense only if reads are specified)\n",
      "    --no-gzip                         Do not compress large output files\n",
      "    --no-read-stats                   Do not align reads to assemblies\n",
      "                                      Reads will be aligned to reference and used for coverage analysis,\n",
      "                                      upper bound assembly simulation, and structural variation detection.\n",
      "                                      Use this option if you do not need read statistics for assemblies.\n",
      "    --fast                            A combination of all speedup options except --no-check\n",
      "\n",
      "Other:\n",
      "    --silent                          Do not print detailed information about each step to stdout (log file is not affected)\n",
      "    --test                            Run QUAST on the data from the test_data folder, output to quast_test_output\n",
      "    --test-sv                         Run QUAST with structural variants detection on the data from the test_data folder,\n",
      "                                      output to quast_test_output\n",
      "-h  --help                            Print full usage message\n",
      "-v  --version                         Print version\n",
      "\n",
      "Online QUAST manual is available at http://quast.sf.net/manual\n"
     ]
    }
   ],
   "source": [
    "!quast.py -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Archivos_sim_ont  Edu_ont_assembly.ipynb  Images  pipeline_notes.ipynb\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/sivico26/anaconda3/lib/python3.6/site-packages/quast-5.0.0.dev0-py3.6.egg/EGG-INFO/scripts/quast.py -t 4 Archivos_sim_ont/Ensamble/fago_lambda_raw.fasta Archivos_sim_ont/Polishing/fago_lambda_pol.fasta -R Archivos_sim_ont/fago_lambda_ref.fasta -o Archivos_sim_ont/quast_results\n",
      "\n",
      "Version: 5.0.0.dev0, 1563de2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.13.0-38-generic-x86_64-with-debian-stretch-sid (linux_64)\n",
      "  Python version: 3.6.3\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2018-04-26 13:57:51\n",
      "\n",
      "Logging to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/quast.log\n",
      "\n",
      "CWD: /home/sivico26/Documents/Notebooks/Pipeline\n",
      "Main parameters: \n",
      "  MODE: default, threads: 4, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: one, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/fago_lambda_ref.fasta ==> fago_lambda_ref\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  1  Archivos_sim_ont/Ensamble/fago_lambda_raw.fasta ==> fago_lambda_raw\n",
      "  2  Archivos_sim_ont/Polishing/fago_lambda_pol.fasta ==> fago_lambda_pol\n",
      "\n",
      "2018-04-26 13:57:52\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    fago_lambda_ref.fasta, length = 48502, num fragments = 1, GC % = 49.86\n",
      "  Contig files: \n",
      "    1  fago_lambda_raw\n",
      "    2  fago_lambda_pol\n",
      "  Calculating N50 and L50...\n",
      "    1  fago_lambda_raw, N50 = 49533, L50 = 1, Total length = 49533, GC % = 49.20, # N's per 100 kbp =  0.00\n",
      "    2  fago_lambda_pol, N50 = 49310, L50 = 1, Total length = 49310, GC % = 49.09, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/basic_stats/GC_content_plot.pdf\n",
      "  Drawing fago_lambda_raw GC content plot...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/basic_stats/fago_lambda_raw_GC_content_plot.pdf\n",
      "  Drawing fago_lambda_pol GC content plot...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/basic_stats/fago_lambda_pol_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2018-04-26 13:57:54\n",
      "Running Contig analyzer...\n",
      "  2  fago_lambda_pol\n",
      "  1  fago_lambda_raw\n",
      "  2  Logging to files /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/contigs_reports/contigs_report_fago_lambda_pol.stdout and contigs_report_fago_lambda_pol.stderr...\n",
      "  1  Logging to files /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/contigs_reports/contigs_report_fago_lambda_raw.stdout and contigs_report_fago_lambda_raw.stderr...\n",
      "  2  Aligning contigs to the reference\n",
      "  1  Aligning contigs to the reference\n",
      "  1  Analysis is finished.\n",
      "  2  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2018-04-26 13:57:57\n",
      "Running NA-NGA calculation...\n",
      "  1  fago_lambda_raw, Largest alignment = 113, NA50 = None, LA50 = None\n",
      "  2  fago_lambda_pol, Largest alignment = 34257, NA50 = 34257, NGA50 = 34257, LA50 = 1, LGA50 = 1\n",
      "  Drawing cumulative plot...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2018-04-26 13:57:58\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  1  fago_lambda_raw\n",
      "  2  fago_lambda_pol\n",
      "  1  Analysis is finished.\n",
      "  2  Analysis is finished.\n",
      "  Drawing Genome fraction, % histogram...\n",
      "    saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/genome_stats/genome_fraction_histogram.pdf\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2018-04-26 13:57:59\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2018-04-26 13:58:01\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/report.html\n",
      "  PDF version (tables and plots) is saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/report.pdf\n",
      "  Icarus (contig browser) is saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/icarus.html\n",
      "  Log saved to /home/sivico26/Documents/Notebooks/Pipeline/Archivos_sim_ont/quast_results/quast.log\n",
      "\n",
      "Finished: 2018-04-26 13:58:01\n",
      "Elapsed time: 0:00:09.711855\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n"
     ]
    }
   ],
   "source": [
    "!quast.py -t 4 Archivos_sim_ont/Ensamble/fago_lambda_raw.fasta Archivos_sim_ont/Polishing/fago_lambda_pol.fasta -R Archivos_sim_ont/fago_lambda_ref.fasta -o Archivos_sim_ont/quast_results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
